### PR TITLE
Add singleMachine disk tracking (resolves #770)

### LINKED ISSUE #770: Add tracking of disk to single machine
Should be straight-forward to implement but it will increase the likelihood of deadlock due to the  nested, blocking nature of resource allocation in SingleMachine.

### PR DESCRIPTION
**Pull request by arkal**
resolves #770

### CODE REVIEW

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
I haven't done a review yet, but write the commit message for 025adaf as if you didn't know that it was going  to fail. Just the describe what the test would prove, assuming they pass. The current commit message is going to confuse people in a few months. Also, the title to should refer to the ticket as `(connected to #…)`

**Comment by arkal** [comment]
okay :)

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 1]
```diff
@@ -352,5 +352,6 @@ def __init__(self, resource, requested, available):
         self.resource = resource
 
     def __str__(self):
-        return 'Requesting more {} than available. Requested: {}, Available: {}' \
-               ''.format(self.resource, self.requested, self.available)
+        return 'Requesting more {} than either physically available, or enforced by --max{}. ' \
+               'Requested: {}, Available: {}'.format(self.resource.capitalize(), self.resource,
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/abstractBatchSystem.py]: shouldn't the 2nd one be capitlized?

Comment (arkal) [inline src/toil/batchSystems/abstractBatchSystem.py]: whoops. I'll fix that.

File: src/toil/batchSystems/singleMachine.py [hunk 2]
```diff
@@ -118,37 +125,51 @@ def worker(self, inputQueue):
...
+                                                          "from command '%s'.", statusCode,
+                                                          jobCommand)
+                                    finally:
+                                        self.runningJobs.pop(jobID)
+                                finally:
+                                    if statusCode is not None and not info.killIntended:
+                                        self.outputQueue.put((jobID, statusCode,
+                                                              time.time() - startTime))
+                except InsufficientSystemResources as err:
+                    # Suppressing the error message and printing this to debug instead.
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: This is obvious.

File: src/toil/batchSystems/singleMachine.py [hunk 3]
```diff
@@ -118,37 +125,51 @@ def worker(self, inputQueue):
...
+                                                log.error("Got exit code %i (indicating failure) "
+                                                          "from command '%s'.", statusCode,
+                                                          jobCommand)
+                                    finally:
+                                        self.runningJobs.pop(jobID)
+                                finally:
+                                    if statusCode is not None and not info.killIntended:
+                                        self.outputQueue.put((jobID, statusCode,
+                                                              time.time() - startTime))
+                except InsufficientSystemResources as err:
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Exceptions always `e`

File: src/toil/batchSystems/singleMachine.py [hunk 4]
```diff
@@ -241,15 +262,27 @@ def __init__(self, startTime, popen, killIntended):
...
     def acquire(self, amount):
         with self.condition:
+            startTime = time.time()
             while amount > self.value:
-                self.condition.wait()
+                if time.time() - startTime >= 5:
+                    # This means the thread timed out waiting for the resource.  We exit the nested
+                    # context managers in worker to prevent blocking of a resource due to
+                    # unavailability of a nested resource request.
+                    raise InsufficientSystemResources(resource=self.resourceType, requested=amount,
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: This is not the right exception type. Create new one, nest it in ResourcePool and call it `AcquisitionTimeoutException`.

Comment (arkal) [inline src/toil/batchSystems/singleMachine.py]: Haha I had created an `InsufficientResourceAvailable` Exception and realized it was almost exactly the same as `InsufficientSystemResources` (in terms of what variables it holds) barring the error message so I thought i'd appropriate this instead.

File: src/toil/test/batchSystems/batchSystemTest.py [hunk 5]
```diff
@@ -577,6 +577,118 @@ class SingleMachineBatchSystemJobTest(hidden.AbstractBatchSystemJobTest):
...
+        But the order of execution should be
+        root > bl, fJ, sJ > sJC > sJCC > fJC
+        since fJC cannot run till bl finishes but sJC and sJCC can(fJC blocked by disk). If the
+        resource acquisition is written properly, then fJC which is scheduled before sJC and sJCC
+        should not block them, and should only run after they finish.
+        """
+        Job.Runner.startToil(root, options)
+        with open(outFile) as oFH:
+            outString = oFH.read()
+        self.assertEqual(outString, 'AAACDB')
```
Comment (hannes-ucsc) [inline src/toil/test/batchSystems/batchSystemTest.py]: Can you make the writeVal values consistent with the job name? That way I won't have to hunt down each job's writeVal to verify your assertion against what you write in the comment above.

File: src/toil/test/batchSystems/batchSystemTest.py [hunk 6]
```diff
@@ -577,6 +577,118 @@ class SingleMachineBatchSystemJobTest(hidden.AbstractBatchSystemJobTest):
...
+                                         disk='1M')
+
+        root.addChild(blocker)
+        root.addChild(firstJob)
+        root.addChild(secondJob)
+
+        firstJob.addChild(firstJobChild)
+        secondJob.addChild(secondJobChild)
+
+        secondJobChild.addChild(secondJobChildChild)
```
Comment (hannes-ucsc) [inline src/toil/test/batchSystems/batchSystemTest.py]: "Grand child"

File: src/toil/test/batchSystems/batchSystemTest.py [hunk 7]
```diff
@@ -577,6 +577,118 @@ class SingleMachineBatchSystemJobTest(hidden.AbstractBatchSystemJobTest):
...
+
+        root = Job()
+        # writes A
+        blocker = Job.wrapFn(_resourceBlockTestAuxFn, outFile=outFile, sleepTime=30, writeVal='A',
+                             cores=2, memory='1M', disk='1M')
+        firstJob = Job.wrapFn(_resourceBlockTestAuxFn, outFile=outFile, sleepTime=5, writeVal='A',
+                              cores=1, memory='1M', disk='1M')
+        secondJob = Job.wrapFn(_resourceBlockTestAuxFn, outFile=outFile, sleepTime=10, writeVal='A',
+                               cores=1, memory='1M', disk='1M')
+        # Writes B
```
Comment (hannes-ucsc) [inline src/toil/test/batchSystems/batchSystemTest.py]: Obvious, but see below.

File: src/toil/batchSystems/singleMachine.py [hunk 8]
```diff
@@ -118,37 +125,51 @@ def worker(self, inputQueue):
...
-                                                  "command '%s'.", statusCode, jobCommand)
-                            finally:
-                                self.runningJobs.pop(jobID)
-                        finally:
-                            if statusCode is not None and not info.killIntended:
-                                self.outputQueue.put((jobID, statusCode, time.time() - startTime))
-            finally:
-                log.debug('Finished job. self.coreFractions ~ %s and self.memory ~ %s',
-                          self.coreFractions.value, self.memory.value)
+            while True:
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Here's what I don't understand: this loops get tries to satisfy the same job with every iteration. How can this be effective in increasing job throughput?

Comment (arkal) [inline src/toil/batchSystems/singleMachine.py]: From what i've understood, each job gets one worker. This loop keeps the current worker alive and doesn't return a value saying the job passed or failed. Another concurrent worker can then attempt to use the resources that this one would have potentially blocked.

File: src/toil/test/batchSystems/batchSystemTest.py (hunk 5 above)
Comment (arkal) [inline src/toil/test/batchSystems/batchSystemTest.py]: Done!

File: src/toil/batchSystems/singleMachine.py (hunk 4 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: The likelihood is incidental.

File: src/toil/batchSystems/singleMachine.py (hunk 8 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Ahh, yes. Now I remember.

File: src/toil/batchSystems/singleMachine.py (hunk 4 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Using a specific exception will increase the resusability of ResourcePool.

File: src/toil/batchSystems/singleMachine.py [hunk 9]
```diff
@@ -241,15 +262,27 @@ def __init__(self, startTime, popen, killIntended):
...
+                    # This means the thread timed out waiting for the resource.  We exit the nested
+                    # context managers in worker to prevent blocking of a resource due to
+                    # unavailability of a nested resource request.
+                    raise InsufficientSystemResources(resource=self.resourceType, requested=amount,
+                                                      available=self.value)
+                # Allow 5 seconds to get the resource, else quit through the above if condition.
+                # This wait + timeout is the last thing in the loop such that a request that takes
+                # longer than 5s due to multiple wakes under the 5 second threshold are still
+                # honored.
+                self.condition.wait(timeout=5)
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: This `5` is related to the other `5` above. It and the sleep of 10 are important parameters that should be promoted to class attributes.

File: src/toil/test/batchSystems/batchSystemTest.py [hunk 10]
```diff
@@ -577,6 +577,118 @@ class SingleMachineBatchSystemJobTest(hidden.AbstractBatchSystemJobTest):
...
+        resource acquisition is written properly, then fJC which is scheduled before sJC and sJCC
+        should not block them, and should only run after they finish.
+        """
+        Job.Runner.startToil(root, options)
+        with open(outFile) as oFH:
+            outString = oFH.read()
+        self.assertEqual(outString, 'AAACDB')
+
+    @staticmethod
+    def _resourceBlockTestAuxFn(outFile, sleepTime, writeVal):
```
Comment (hannes-ucsc) [inline src/toil/test/batchSystems/batchSystemTest.py]: Considering that this is the end the testcase, can we forgo the "export static method as global" hack?

File: src/toil/batchSystems/singleMachine.py (hunk 9 above)
Comment (arkal) [inline src/toil/batchSystems/singleMachine.py]:
```
self.resourceAcquisitionTimeout = 5
self.resourceAcquisitionRetryDelay = 10
```
sound good?

Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: Considering that this is a *Resource*Pool, I would drop the `resource` prefix.

Comment (arkal) [inline src/toil/batchSystems/singleMachine.py]: okai

File: src/toil/test/batchSystems/batchSystemTest.py (hunk 10 above)
Comment (arkal) [inline src/toil/test/batchSystems/batchSystemTest.py]: so just make it a global? that's fine by me!

File: src/toil/batchSystems/singleMachine.py (hunk 4 above)
Comment (arkal) [inline src/toil/batchSystems/singleMachine.py]: sounds good

File: src/toil/test/batchSystems/batchSystemTest.py (hunk 10 above)
Comment (hannes-ucsc) [inline src/toil/test/batchSystems/batchSystemTest.py]: yes, as long as it is in close proximity with the test method.

Comment (arkal) [inline src/toil/test/batchSystems/batchSystemTest.py]: done

**Comment by arkal** [comment]
Tests were timing out due to insufficient disk during setup of single machine batch system for tests. Wasn't a problem before because we didn't track disk haha.

**Comment by arkal** [comment]
@hannes-ucsc this is ready!

File: src/toil/batchSystems/singleMachine.py [hunk 11]
```diff
@@ -275,3 +319,26 @@ def acquisitionOf(self, amount):
             yield
         finally:
             self.release(amount)
+
+
+class AcquisitionTimeoutException(Exception):
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/singleMachine.py]: This should be nested in ResourcePool or renamed to ResourceAcquisitionTimeoutException. I'd prefer the former.

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes

**Pull request merged**